sklearn/metrics/_classification.py

```diff
@@ -463,6 +464,23 @@ def confusion_matrix(
     >>> (tn, fp, fn, tp)
     (0, 2, 1, 1)
     """
+    xp, _, device_ = get_namespace_and_device(y_true, y_pred, labels, sample_weight)
+    y_true = check_array(
+        y_true,
+        dtype=None,
+        ensure_2d=False,
+        ensure_all_finite=False,
+        ensure_min_samples=0,
+    )
+    y_pred = check_array(
+        y_pred,
+        dtype=None,
+        ensure_2d=False,
+        ensure_all_finite=False,
+        ensure_min_samples=0,
+    )
+    y_true = _convert_to_numpy(y_true, xp)
```

Comment:
Since the use of `_convert_to_numpy` should be the exception rather than the rule in the scikit-learn codebase, I think we should at least put a comment to make our motivations explicit.

```suggestion
    # Convert the input arrays to NumPy (on CPU) irrespective of the original
    # namespace and device so as to be able to leverage the the efficient
    # counting operations implemented by SciPy in the coo_matrix constructor.
    # The final results will be converted back to the input namespace and device
    # for the sake of consistency with other metric functions with array API support.
    y_true = _convert_to_numpy(y_true, xp)
```